File: mathsys/dev/typescript/nonterminal.ts
//^
//^ HEAD
//^

//> HEAD -> DATA
import {Binary, Pointer} from "./ir";


//^
//^ NONTERMINAL
//^

//> NONTERMINAL -> START
export abstract class NonTerminal {
    private frozen: boolean = false;
    constructor() {return new Proxy(this, {set: (target, property, value) => {
        if (target.frozen) {throw new Error("Instance is frozen")}
        // @ts-ignore
        target[property] = value;
        return true;
    }})}
    abstract latex(types: Map<string, string>): string;
    abstract ir(binary: Binary[], nodes: Binary[]): Pointer;
    freeze(): void {this.frozen = true}
}

File: mathsys/dev/python/nonterminal.py
#^
#^  HEAD
#^

#> HEAD -> MODULES
from abc import ABC, abstractmethod
from dataclasses import FrozenInstanceError
from typing import Any

#> HEAD -> DATA
from .ir import Binary, Pointer


#^
#^  NONTERMINAL
#^

#> NONTERMINAL -> START
class NonTerminal(ABC):
    frozen: bool = False
    @abstractmethod
    def __init__(self, items: list) -> None: pass
    @abstractmethod
    def latex(self, types: dict[str, str]) -> str: pass
    @abstractmethod
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer: pass
    def freeze(self) -> None: self.frozen = True
    @classmethod
    def placeholder(cls) -> bool: return cls.__class__.__name__.startswith("Level")
    def __setattr__(self, name: str, value: Any) -> None:
        if self.frozen: raise FrozenInstanceError()
        return super().__setattr__(name, value)